### PR TITLE
[Feature] Add AnalogNoise interface

### PR DESCRIPTION
**Pull request by chMoussa**
Solves #299 by adding a torch module `AnalogNoise` that keeps the noise operators, and can expend them to match the qubit support of `HamiltonianEvolution` when using mesolve. Also differentiate between digital noise operations.

### CODE REVIEW

File: docs/noise.md [hunk 1]
```diff
@@ -196,6 +208,6 @@ noiseless_expectation = pyq.expectation(circ, state, {"theta": theta}, observabl
 readobj = ReadoutNoise(n_qubits, seed=0)
 noisycirc = pyq.QuantumCircuit(n_qubits, ops, readobj)
 noisy_expectation = pyq.expectation(noisycirc, state, {"theta": theta}, observable=obs, n_shots=1000)
-print("Noiseless expectation ", noiseless_expectation.item())
-print("Noisy expectation ", noisy_expectation.item())
+print("Noiseless expectation ", noiseless_expectation.item()) # markdown-exec: hide
+print("Noisy expectation ", noisy_expectation.item()) # markdown-exec: hide
```
Comment (RolandMacDoland) [inline docs/noise.md:212]:
```suggestion
print(f"Noiseless expectation: {noiseless_expectation.item()}") # markdown-exec: hide
print(f"Noisy expectation: {noisy_expectation.item()}") # markdown-exec: hide
```

File: pyproject.toml [hunk 2]
```diff
@@ -6,7 +6,7 @@ build-backend = "hatchling.build"
 name = "pyqtorch"
 description = "An efficient, large-scale emulator designed for quantum machine learning, seamlessly integrated with a PyTorch backend. Please refer to https://pyqtorch.readthedocs.io/en/latest/ for setup and usage info, along with the full documentation."
 readme = "README.md"
-version = "1.6.0"
+version = "1.6.1"
```
Comment (RolandMacDoland) [inline pyproject.toml:9]: I guess it feels like a feature enough to bump the minor ?

File: pyqtorch/hamiltonians/evolution.py [hunk 3]
```diff
@@ -260,7 +258,17 @@ def __init__(
         self._cache_hamiltonian_evo: dict[str, Tensor] = dict()
         self.cache_length = cache_length
 
-        self.noise_operators: list[Tensor] = noise_operators
+        if isinstance(noise, list):
+            noise = AnalogNoise(noise, self.qubit_support)
+        if (
+            noise is not None
+            and len(set(noise.qubit_support) - set(self.qubit_support)) > 0
```
Comment (RolandMacDoland) [inline pyqtorch/hamiltonians/evolution.py:265]: Couldn't you simply compare supports ?

File: pyqtorch/noise/analog.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    SolverType,
+    density_mat,
+    expand_operator,
+    permute_basis,
+)
+
+
+class AnalogNoise(torch.nn.Module):
+    """AnalogNoise is used within `HamiltonianEvolution`
+        when using a Shrodinger equation solver.
```
Comment (RolandMacDoland) [inline pyqtorch/noise/analog.py:23]:
```suggestion
    """AnalogNoise is used within `HamiltonianEvolution`
        when solving the Schrödinger or a master (Lindblad) equation.
```

File: pyqtorch/noise/analog.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        self._qubit_support = (
+            qubit_support
+            if isinstance(qubit_support, Support)
+            else Support(target=qubit_support)
+        )
+
+        for index, tensor in enumerate(noise_operators):
+            if len(self._qubit_support) < int(log(tensor.shape[1], 2)):
+                raise ValueError(
+                    "Tensor dimensions should match the length of the qubit support."
```
Comment (RolandMacDoland) [inline pyqtorch/noise/analog.py:49]:
```suggestion
                    f"Tensor dimensions should match the length of the qubit support. Tensor {tensor} has incompatible dimensions {int(log(tensor.shape[1], 2)}."
```

File: pyqtorch/noise/analog.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+
+class AnalogDepolarizing(AnalogNoise):
+    """
+    Defines jump operators for an Analog Depolarizing channel.
+
+    Under the depolarizing noise, a system in any state evolves
+      to the maximally mixed state at a rate :math:`p`.
+
+    The corresponding jump operatorsare :
```
Comment (RolandMacDoland) [inline pyqtorch/noise/analog.py:172]:
```suggestion
    The corresponding jump operators are :
```

File: pyqtorch/qubit_support.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    """
+
+    def __init__(
+        self,
+        target: int | tuple[int, ...],
+        control: int | tuple[int, ...] | None = None,
+    ) -> None:
+        self.target = qubit_support_as_tuple(target)
+        self.control = qubit_support_as_tuple(control) if control is not None else ()
+        # if self.qubits != tuple(set(self.qubits)):
```
Comment (RolandMacDoland) [inline pyqtorch/qubit_support.py:30]: Remove ?

File: tests/test_noise.py [hunk 8]
```diff
@@ -360,3 +361,21 @@ def test_param_noise_apply(
         psi_expected = op_concrete(psi_init, values=values)
         psi_star = op_concrete_noise(psi_init, values=values)
         assert torch.allclose(psi_star, psi_expected, rtol=RTOL, atol=ATOL)
+
+
+def test_analog_noise_add():
```
Comment (RolandMacDoland) [inline tests/test_noise.py:366]: Isn't there some known example of analog noisy simulations that we can compare with ?

**Review by RolandMacDoland — APPROVED** [review]
Thanks a lot @chMoussa. LGTM. Just wondering if we can compare noisy simulations with some known case to get concrete numbers to check on.

File: pyqtorch/hamiltonians/evolution.py (hunk 3 above)
Comment (chMoussa) [reply pyqtorch/hamiltonians/evolution.py:265]: Sure, removing len

File: tests/test_noise.py (hunk 8 above)
Comment (chMoussa) [reply tests/test_noise.py:366]: Hm could not find any, usually this is compared to mesolve which is already tested.

**Pull request merged**